### PR TITLE
Upgrader - platform specific directory creation

### PR DESCRIPTION
**Pull request by jamill**
The upgrade logic between the different platforms has some high level differences that need to be accounted for:

1) Creation of directories with the correct ACLs
    - On Mac, the directories will be created with the expected ACLs, and we will not need to re-acl the directories.

There are other ways to handle the platform differences, but including:

  - performing the re-ACLing of directories on non-windows platforms. We could re-use more of the top level logic, but then we would be doing extra work with ACLs that is not needed.

  - Do no re-ACL on non-windows platforms. I did not want to cause confusion with the methods not doing what they advertise (especially around ACLs), in case someone misses this subtlety in the future. This would be hard from a maintainability standpoint...

2) Whether the Upgrader is responsible for remounting or not
    - On Windows, the upgrader itself is responsible for re-mounting after upgrade, while on Mac, the GVFS.Service will auto remount when the service starts up.

There is still some further work I need to do here to get this into shape, but I wanted to put this up for initial feedback on the general direction.

### CODE REVIEW

File: GVFS/GVFS.Common/ProductUpgraderPlatformStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public abstract class ProductUpgraderPlatformStrategy
+    {
+        public ProductUpgraderPlatformStrategy(PhysicalFileSystem fileSystem, ITracer tracer)
+        {
+            this.FileSystem = fileSystem;
+            this.Tracer = tracer;
+        }
+
+        protected PhysicalFileSystem FileSystem { get; private set; }
```
Comment (derrickstolee) [inline GVFS/GVFS.Common/ProductUpgraderPlatformStrategy.cs]: nit: drop the `private set;` and it will become read-only (only assignable in the constructor).

File: GVFS/GVFS.Common/ProductUpgraderPlatformStrategyFactory.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public static ProductUpgraderPlatformStrategy Create(
+            PhysicalFileSystem fileSystem,
+            ITracer tracer)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                return new MacProductUpgraderPlatformStrategy(fileSystem, tracer);
+            }
+            else
```
Comment (derrickstolee) [inline GVFS/GVFS.Common/ProductUpgraderPlatformStrategyFactory.cs]: Perhaps be specific here and have the `else` condition throw a `NotSupportedException` or something? That way the Linux platform can slot in easilyi.

File: GVFS/GVFS.Common/WindowsProductUpgraderPlatformStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            string rootDirectoryPath = ProductUpgraderInfo.GetUpgradesDirectoryPath();
+            string toolsDirectoryPath = Path.Combine(rootDirectoryPath, ProductUpgrader.ToolsDirectory);
+
+            Exception deleteDirectoryException;
+            if (this.FileSystem.DirectoryExists(toolsDirectoryPath) &&
+                !this.FileSystem.TryDeleteDirectory(toolsDirectoryPath, out deleteDirectoryException))
+            {
+                error = $"Failed to delete {toolsDirectoryPath} - {deleteDirectoryException.Message}";
+
+                /// this.TraceException(deleteDirectoryException, nameof(this.TrySetupToolsDirectory), $"Error deleting {toolsDirectoryPath}.");
```
Comment (derrickstolee) [inline GVFS/GVFS.Common/WindowsProductUpgraderPlatformStrategy.cs]: Is this intended to be commented-out?

File: GVFS/GVFS.Upgrader/MacUpgradeOrchestrator.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+using CommandLine;
```
Comment (derrickstolee) [inline GVFS/GVFS.Upgrader/MacUpgradeOrchestrator.cs]: nit: many of these `using`s seem unused.

File: GVFS/GVFS.Upgrader/Program.cs [hunk 5]
```diff
@@ -9,8 +9,18 @@ public static void Main(string[] args)
         {
             GVFSPlatformLoader.Initialize();
 
-            Parser.Default.ParseArguments<UpgradeOrchestrator>(args)
+            UpgradeOrchestrator upgraderType = UpgradeOrchestratorFactory.Create();
+
+            if (upgraderType is MacUpgradeOrchestrator)
```
Comment (derrickstolee) [inline GVFS/GVFS.Upgrader/Program.cs]: This whole chunk is confusing to me.

1. We should use our base type to avoid this `is` convention.
2. We create `upgraderType` just to check the `is`, and then we don't even use it.

It would be better if the `UpgradeOrchestrator` had some `ExecuteOnArgs(string [] args)` method that did this internal logic according to its own type.

File: GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs [hunk 6]
```diff
@@ -152,6 +152,35 @@ public void Execute()
             Environment.ExitCode = (int)this.ExitCode;
         }
 
+        protected virtual bool TryMountRepositories(out string consoleError)
```
Comment (derrickstolee) [inline GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs]: Should this be `abstract` in `UpgradeOrchestrator` and the implementation actually live in a `WindowsUpgradeOrchestrator`?

File: GVFS/GVFS.Upgrader/Program.cs (hunk 5 above)
Comment (derrickstolee) [reply GVFS/GVFS.Upgrader/Program.cs]: I see now that there is no `WindowsUpgradeOrchestrator`, which would probably be helpful here.

File: GVFS/GVFS.Upgrader/UpgradeOrchestratorFactory.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+namespace GVFS.Upgrader
+{
+    public static class UpgradeOrchestratorFactory
+    {
+        public static UpgradeOrchestrator Create()
+        {
+#if MACOS_BUILD
+            return new MacUpgradeOrchestrator();
+#else
```
Comment (derrickstolee) [inline GVFS/GVFS.Upgrader/UpgradeOrchestratorFactory.cs]: `#elseif WINDOWS_BUILD` to avoid issues with Linux platform later.

File: GVFS/GVFS.Common/MacProductUpgraderPlatformStrategy.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace GVFS.Common
+{
+    public class MacProductUpgraderPlatformStrategy : ProductUpgraderPlatformStrategy
```
Comment (kewillford) [inline GVFS/GVFS.Common/MacProductUpgraderPlatformStrategy.cs]: Still looking but I'm not sure what `PlatformStrategy` means.  

I also don't think we should have platform specific code going into `GVFS.Common`.  Looking at this and it seems these should be a part of `IPlatformFileSystem` with a `TryPrepareDirectory(string path, out string error)`.  We might even be able to remove the 2 permission method from that interface since they appear only to be for Windows.
```
        bool TryCreateDirectoryWithAdminAndUserModifyPermissions(string directoryPath, out string error);
        bool TryCreateOrUpdateDirectoryToAdminModifyPermissions(ITracer tracer, string directoryPath, out string error);
```

File: GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj [hunk 9]
```diff
@@ -26,6 +26,7 @@
     <When Condition="'$(OS)' == 'Windows_NT'">
       <PropertyGroup>
         <TargetFrameworks>net461;netcoreapp2.1</TargetFrameworks>
+	<DefineConstants>$(DefineConstants);WINDOWS_BUILD</DefineConstants>
```
Comment (kewillford) [inline GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj]: Looks like this project is already referencing the Windows or the Mac Platform projects so why can't we have the platform specific code live in those projects instead of adding another way to create platform specific code?

File: GVFS/GVFS.Common/MacProductUpgraderPlatformStrategy.cs (hunk 8 above)
Comment (jamill) [reply GVFS/GVFS.Common/MacProductUpgraderPlatformStrategy.cs]: > it seems these should be a part of IPlatformFileSystem with a TryPrepareDirectory(string path, out string error)

I like this idea - but the problem I ran into is that that action depends on the directory (which the upgrader has the most context on).

File: GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj (hunk 9 above)
Comment (jamill) [reply GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj]: That is an option - but then we are putting upgrader specific code in the platform project...

Comment (jamill) [reply GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj]: I will look at moving this into the platform projects

Comment (jamill) [reply GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj]: I will look at moving this into the platform projects

Comment (jamill) [reply GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj]: I will look at moving this into platform specific projects

File: GVFS/GVFS.Common/ProductUpgraderPlatformStrategyFactory.cs (hunk 2 above)
Comment (jamill) [reply GVFS/GVFS.Common/ProductUpgraderPlatformStrategyFactory.cs]: I liked this suggestion - but moved the "Creation" logic into the Platform classes, so that should help with the Linux platform.

File: GVFS/GVFS.Common/WindowsProductUpgraderPlatformStrategy.cs (hunk 3 above)
Comment (jamill) [reply GVFS/GVFS.Common/WindowsProductUpgraderPlatformStrategy.cs]: I was lazy for my initial round of changes. The base class did not have the `TraceException` functionality. I added this to the base class, and uncommented this (and similar) lines.

File: GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj (hunk 9 above)
Comment (jamill) [reply GVFS/GVFS.Upgrader/GVFS.Upgrader.csproj]: After thinking about this a bit more - I moved the platform specific code out of the Common project, but I am still am not sure about the `GVFS.Upgrader` project. This project is almost a separate application, and I didn't want to bring high level "other application" specific logic into the Platform classes.

I am open to thoughts on this, but I wasn't going to convert this just yet.

File: GVFS/GVFS.Common/MacProductUpgraderPlatformStrategy.cs (hunk 8 above)
Comment (jamill) [reply GVFS/GVFS.Common/MacProductUpgraderPlatformStrategy.cs]: If there is an appropriate abstraction here, I am open to adding these directly to the `IPlatformFileSystem` class, but I haven't figured out a non-upgrader specific abstraction here... I will continue to think about this.

**Comment by jamill** [comment]
Thanks @derrickstolee for the review!

> This means having an abstract base class with an implementation for each platform.

I think this makes sense as a goal. The problem I ran into is that we have a bunch of existing tests that exercise the current concrete implementation. If we make the `UpgradeOrchestrator` class abstract, then these tests will need reaction work... Either a "Testable" `UpgradeOrchestrator` class to test the common logic (which is most of the class), moving the tests to follow the new concrete implementations, etc... I like the suggestion - so let me investigate it a bit further.

> It would be nice to do some research into the Linux platform a bit here so we can figure out what can be combined into a POSIX platform class 

Good idea! I think a lot (or at least some) of the logic that is in the Mac specific class can actually go in the POSIX layer instead (as you noted).

File: GVFS/GVFS.Common/GVFSPlatform.cs [hunk 10]
```diff
@@ -73,6 +73,10 @@ public static void Register(GVFSPlatform platform)
             ITracer tracer,
             string lockPath);
 
+        public abstract ProductUpgraderPlatformStrategy CreateProductUpgraderPlatformInteractions(
```
Comment (jamill) [inline GVFS/GVFS.Common/GVFSPlatform.cs]: I think the `ProductUpgraderPlatformStrategy` name is overly verbose and at the same time not clear. I don't think the `Interactions` suffix is any better... Still thinking for a better name.

File: GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs (hunk 6 above)
Comment (jamill) [reply GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs]: Yes - ideally this should be the case. For the current state of this, this interacts with how the tests are architected. I am still looking into what it would take to move to this structure (in which case most of the logic would remain the same, but the implementation would slide down into a Windows specific type). If it introduces a lot of churn, I was going to defer this (and not the followup work item)

File: GVFS/GVFS.Upgrader/Program.cs (hunk 5 above)
Comment (jamill) [reply GVFS/GVFS.Upgrader/Program.cs]: I was able to clean this up a bit. I now have a stable `UpgradeOptions` class that then gets passed to a specific `UpgradeOrchestrator` implementation.

File: GVFS/GVFS.Upgrader/WindowsUpgradeOrchestrator.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            : base(upgrader, tracer, fileSystem, preRunChecker, input, output)
+        {
+        }
+
+        public WindowsUpgradeOrchestrator(UpgradeOptions options)
+            : base(options)
+        {
+        }
+
+        protected override bool TryMountRepositories(out string consoleError)
```
Comment (derrickstolee) [inline GVFS/GVFS.Upgrader/WindowsUpgradeOrchestrator.cs]: Should this be in the base (as `virtual`) and then the Mac version can leave it unimplemented until it is supported?

File: GVFS/GVFS.Upgrader/UpgradeOptions.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
+using CommandLine;
+
+namespace GVFS.Upgrader
+{
+    [Verb("UpgradeOrchestrator", HelpText = "Upgrade VFS for Git.")]
```
Comment (derrickstolee) [inline GVFS/GVFS.Upgrader/UpgradeOptions.cs]: This `Verb` annotation seems out of place.

Comment (jamill) [reply GVFS/GVFS.Upgrader/UpgradeOptions.cs]: This usage is different than how we generally structure this in our product, but it does match the intended usage:

https://github.com/commandlineparser/commandline#for-verbs

I think we could ask whether we need this to be a verb, as this is being used to process command line arguments for a program that doesn't actually have verbs. I will note this as a potential cleanup.

File: GVFS/GVFS.Common/ProductUpgrader.cs [hunk 13]
```diff
@@ -154,20 +173,7 @@ public virtual bool TrySetupToolsDirectory(out string upgraderToolPath, out stri
...
-                return false;
-            }
-
-            if (!this.fileSystem.TryCreateOrUpdateDirectoryToAdminModifyPermissions(
-                    this.tracer,
-                    toolsDirectoryPath,
-                    out error))
+            if (!this.productUpgraderPlatformStrategy.TryPrepareApplicationDirectory(out error))
             {
                 upgraderToolPath = null;
```
Comment (alameenshah) [inline GVFS/GVFS.Common/ProductUpgrader.cs]: Not directly related to this PR. Can you take a look at lines #182 through #186. `TrySetupToolsDirectory` is trying to copy everything inside `C:\Program Files\GVFS` to "C:\ProgramData\GVFS\GVFS.Upgrade\Tools". The whole set of `GVFS` assemblies get copied. This is not the intent (we only need GVFS.Upgrade.exe and its dependencies), right?

File: GVFS/GVFS.Common/ProductUpgraderPlatformStrategy.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        protected PhysicalFileSystem FileSystem { get; }
+        protected ITracer Tracer { get; }
+
+        public abstract bool TryPrepareLogDirectory(out string error);
+
+        public abstract bool TryPrepareApplicationDirectory(out string error);
+
+        public abstract bool TryPrepareDownloadDirectory(out string error);
+
+        protected void TraceException(Exception exception, string method, string message)
```
Comment (alameenshah) [inline GVFS/GVFS.Common/ProductUpgraderPlatformStrategy.cs]: This is a useful utility method, that is not related to `ProductUpgraderPlatformStrategy`. Thoughts on moving this to `JsonTracer` as a new `RelatedException` method?

File: GVFS/GVFS.Platform.Mac/MacProductUpgraderPlatformStrategy.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+            if (this.FileSystem.DirectoryExists(toolsDirectoryPath) &&
+                !this.FileSystem.TryDeleteDirectory(toolsDirectoryPath, out deleteDirectoryException))
+            {
+                error = $"Failed to delete {toolsDirectoryPath} - {deleteDirectoryException.Message}";
+
+                this.TraceException(deleteDirectoryException, nameof(this.TryPrepareApplicationDirectory), $"Error deleting {toolsDirectoryPath}.");
+                return false;
+            }
+
+            this.FileSystem.CreateDirectory(toolsDirectoryPath);
```
Comment (alameenshah) [inline GVFS/GVFS.Platform.Mac/MacProductUpgraderPlatformStrategy.cs]: Upgrade verb runs elevated on the Mac too, right? Would this end up creating a root:admin owned directory inside user's `$HOME/Application\ Support/GVFS`? (root owned directory inside user home will be unusual.)

File: GVFS/GVFS.Common/ProductUpgraderPlatformStrategy.cs (hunk 14 above)
Comment (jamill) [reply GVFS/GVFS.Common/ProductUpgraderPlatformStrategy.cs]: Yeah - good point. We duplicate this utility method is several places, and we should probably centralize it. I will include it as a potential cleanup (and standardize this everywhere).

File: GVFS/GVFS.Platform.Mac/MacProductUpgraderPlatformStrategy.cs (hunk 15 above)
Comment (jamill) [reply GVFS/GVFS.Platform.Mac/MacProductUpgraderPlatformStrategy.cs]: > Would this end up creating a root:admin owned directory inside user's $HOME/Application\ Support/GVFS?

Good question - the "application" directory will be created under /usr/local/vfsforgit_upgrader, which is root owned.

File: GVFS/GVFS.Common/ProductUpgrader.cs (hunk 13 above)
Comment (jamill) [reply GVFS/GVFS.Common/ProductUpgrader.cs]: > The whole set of GVFS assemblies get copied. This is not the intent (we only need GVFS.Upgrade.exe and its dependencies), right?

The problem was that it was fragile to keep the list of dependencies straight (across all platforms). We now copy over the entire "application", including some extra files, but I don't think there should be an issue with copying over these extra file... If there are concerns with copying the extra files, please let me know :)

Comment (alameenshah) [reply GVFS/GVFS.Common/ProductUpgrader.cs]: I see...
